lib/vtls/gtls.c

```diff
@@ -1700,20 +1710,37 @@ gtls_connect_common(struct Curl_cfilter *cf,
       result = rc;
       goto out;
     }
+    backend = (struct gtls_ssl_backend_data *)
+                  connssl->backend;
+    session = backend->gtls.session;
+
+    if(data->set.ssl.earlydata &&
+        gnutls_protocol_get_version(session) == GNUTLS_TLS1_3 &&
+        connssl->reused_session &&
+        gnutls_record_get_max_early_data_size(session)) {
+      /* the user wants to send early data
+       * finish the connection in gtls_send
+       */
+      infof(data, "Sending early data");
+      *done = TRUE;
+      connssl->state = ssl_connection_deferred;
+      connssl->connecting_state = ssl_connect_1_epending;
+      return result;
+    }
   }
 
-  rc = handshake(cf, data, TRUE, nonblocking);
-  if(rc) {
-    /* handshake() sets its own error message with failf() */
-    result = rc;
-    goto out;
+  if(ssl_connect_2 == connssl->connecting_state) {
+    rc = handshake(cf, data, TRUE, nonblocking);
+    if(rc) {
+      /* handshake() sets its own error message with failf() */
+      result = rc;
+      goto out;
+    }
   }
 
   /* Finish connecting once the handshake is done */
-  if(ssl_connect_1 == connssl->connecting_state) {
-    struct gtls_ssl_backend_data *backend =
-      (struct gtls_ssl_backend_data *)connssl->backend;
-    gnutls_session_t session;
+  if(ssl_connect_3 == connssl->connecting_state) {
+    backend = (struct gtls_ssl_backend_data *)connssl->backend;
     DEBUGASSERT(backend);
```

Comment:
This assert probably needs to be moved to immediately after the first assigning of this variable.